### PR TITLE
[WIP] contrib-ads 6 post

### PR DESCRIPTION
**Pull request by incompl**

### CODE REVIEW

**Comment by gkatsev** [comment]
Deploy preview for *videojs-blog* ready!

Built with commit f1fe561fbc8bc3e8315484318f92d3a664bb63eb

https://deploy-preview-26--videojs-blog.netlify.com

File: source/_drafts/videojs-contrib-ads-6.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+## How it works
+
+For a history of how the state machine has evolved over time, [there is detailed information in this Github issue](https://github.com/videojs/videojs-contrib-ads/issues/320). In this article we're going to focus on how it works in version 6.
+
+Here is a diagram of the current states:
+
+![](./ad-states.png)
+
+There are two types of states: content states (blue) and ad states (yellow). `player.ads.isInAdMode()` is implemented using this distinction. [Ad mode](https://github.com/videojs/videojs-contrib-ads#ad-mode-definition) is defined as _if content playback is blocked by the ad plugin_. This does not necessarily mean that an ad is playing. For example, if there is a preroll ad, after a play request we show a loading spinner until ad playback actually begins. This is considered part of ad mode.
```
Comment (gkatsev) [inline source/_drafts/videojs-contrib-ads-6.md]: should we make clear that ad states are the states when ad mode is true or something like that? For me, it's a bit abrupt to go from "two types of states" to "ad mode"

File: source/_drafts/videojs-contrib-ads-6.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+Now we're in the ContentPlayback state. At this point, the ad plugin could play a midroll ad, causing a brief foray into the Midroll state, or content could continue without interruption.
+
+When content ends, normally we would see an `ended` event right away. Instead, contrib-ads sends a `contentended` event, which is the ad plugin's chance to play postroll ads. Perhaps the ad plugin has decided not to play postroll ad, so it responds by triggering `nopreroll`. contrib-ads knows that now we're really done, so an `ended` event is triggered.
+
+Finally, since content has ended for the first time, we now transition to the AdsDone state. There will be no more ads, even if the use seeks backwards and plays through the content again.
```
Comment (gkatsev) [inline source/_drafts/videojs-contrib-ads-6.md]: typo: `use` seeks backwards -> `user` seeks backwards

File: source/_drafts/videojs-contrib-ads-6.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
+---
+title: videojs-contrib-ads 6
```
Comment (ldayananda) [inline source/_drafts/videojs-contrib-ads-6.md]: Maybe "videojs-contrib-ads v6 Released"?

File: source/_drafts/videojs-contrib-ads-6.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  name: Greg Smith
+  github: incompl
+tags:
+---
+
+A major update to videojs-contrib-ads has been released. In this post, we'll take a look at what videojs-contrib-ads offers and what has improved in this latest version.
+
+## What is contrib-ads?
+
+Videojs-contrib-ads is a framework for creating Video.js ad plugins. Seamlessly integrating ad support into a video player can be a daunting task, especially if you have other plugins that may be effected. For example, contrib-ads has a feature called [Redispatch](https://github.com/videojs/videojs-contrib-ads#redispatch) that makes sure [media events](https://developer.mozilla.org/en-US/docs/Web/Guide/Events/Media_events) are not impacted by the presence of ads. You may have some JavaScript on your site, or another Video.js plugin, that responds to events such as the `play` event. With contrib-ads, if an ad plays, it will not send an additional `play` event. This keeps things simple as possible without breaking other plugins. [Additional benefits of using contrib-ads are listed in the readme](https://github.com/videojs/videojs-contrib-ads#benefits).
```
Comment (ldayananda) [inline source/_drafts/videojs-contrib-ads-6.md]: I would rephrase the redispatch part by explaining the problem first, then mentioning the contrib-ads feature.

File: source/_drafts/videojs-contrib-ads-6.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+Version 6 of contrib-ads is a maintenance release that includes a major code refactor. The core of the project's behavior is handled by a state machine that advances through various states as the player plays content, preroll ads, midroll ads, and postroll ads. The initial state machine was designed to implement a specific set of functionality when the project was first conceptualized. Over the years, the project became much more fully-featured and many new scenarios and edge-cases arose. In this refactor, the state machine was updated to precisely match the modern feature-set. This has resulted in a more maintainable and reliable codebase.
+
+## How it works
+
+For a history of how the state machine has evolved over time, [there is detailed information in this Github issue](https://github.com/videojs/videojs-contrib-ads/issues/320). In this article we're going to focus on how it works in version 6.
+
+Here is a diagram of the current states:
+
+![](./ad-states.png)
```
Comment (ldayananda) [inline source/_drafts/videojs-contrib-ads-6.md]: no link?

File: source/_drafts/videojs-contrib-ads-6.md (hunk 1 above)
Comment (ldayananda) [inline source/_drafts/videojs-contrib-ads-6.md]: I would explain what ad mode is first before mentioning `isInAdMode`

Comment (ldayananda) [reply source/_drafts/videojs-contrib-ads-6.md]: For example, if there is a preroll ad, after a play request we show a loading spinner until **an ad plugin starts an ad** . This is considered part of ad mode.

Comment (ldayananda) [reply source/_drafts/videojs-contrib-ads-6.md]: Also, a distinction in terminology between contrib-ads and an integrating ad plugin should be made before we get here 😄

File: source/_drafts/videojs-contrib-ads-6.md [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+Now, the user clicks play. At this point ad mode begins and contrib-ads moves forward to the Preroll state. The Preroll state shows a loading spinner while ads load. Once the ad break begins and an ad plays, the control bar turns yellow and will not allow seeking. If the ad takes too long to begin playing, a timeout will occur and content will resume without further adieu. When the ad is complete, contrib-ads restores the content video. Only when content playback results in a `playing` event does ad mode end.
+
+Now we're in the ContentPlayback state. At this point, the ad plugin could play a midroll ad, causing a brief foray into the Midroll state, or content could continue without interruption.
+
+When content ends, normally we would see an `ended` event right away. Instead, contrib-ads sends a `contentended` event, which is the ad plugin's chance to play postroll ads. Perhaps the ad plugin has decided not to play postroll ad, so it responds by triggering `nopreroll`. contrib-ads knows that now we're really done, so an `ended` event is triggered.
```
Comment (ldayananda) [inline source/_drafts/videojs-contrib-ads-6.md]: so it responds by triggering ~~`nopreroll`~~ `nopostroll`

File: source/_drafts/videojs-contrib-ads-6.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+Finally, since content has ended for the first time, we now transition to the AdsDone state. There will be no more ads, even if the use seeks backwards and plays through the content again.
+
+## Conclusion
+
+Ads may not be your favorite part of watching video online, but contrib-ads makes sure that they don't break your site and that content plays even if ads fail. If you're interested in learning more or contributing to the project, [check it out on Github](https://github.com/videojs/videojs-contrib-ads)!
```
Comment (ldayananda) [inline source/_drafts/videojs-contrib-ads-6.md]: >Ads may not be your favorite part of watching video online

😂

File: source/_drafts/videojs-contrib-ads-6.md (hunk 3 above)
Comment (incompl) [reply source/_drafts/videojs-contrib-ads-6.md]: I think "videojs-contrib-ads 6" is a bit more evergreen: this post will still be about what changed in contrib-ads 6 even after it's not "news"

File: source/_drafts/videojs-contrib-ads-6.md (hunk 5 above)
Comment (incompl) [reply source/_drafts/videojs-contrib-ads-6.md]: That is where alt text goes, not a link, if I understand the question correctly.

File: source/_drafts/videojs-contrib-ads-6.md (hunk 1 above)
Comment (incompl) [reply source/_drafts/videojs-contrib-ads-6.md]: Made an update, how does that look?

File: source/_drafts/videojs-contrib-ads-6.md (hunk 2 above)
Comment (incompl) [reply source/_drafts/videojs-contrib-ads-6.md]: Fixed

File: source/_drafts/videojs-contrib-ads-6.md (hunk 4 above)
Comment (incompl) [reply source/_drafts/videojs-contrib-ads-6.md]: Made a change, how does that look?

File: source/_drafts/videojs-contrib-ads-6.md (hunk 6 above)
Comment (incompl) [reply source/_drafts/videojs-contrib-ads-6.md]: Fixed

File: source/_drafts/videojs-contrib-ads-6.md (hunk 5 above)
Comment (gkatsev) [reply source/_drafts/videojs-contrib-ads-6.md]: our template will put this text as a title for the image as well:
```md
![IE8](IE8-small.png)
```
screen shot 2018-03-09 at 17 17 34

File: source/_drafts/videojs-contrib-ads-6.md (hunk 1 above)
Comment (incompl) [reply source/_drafts/videojs-contrib-ads-6.md]: I restructured this section, does that seem clearer?

File: source/_drafts/videojs-contrib-ads-6.md [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+## Version 6
+
+Version 6 of contrib-ads is a maintenance release that includes a major code refactor. The core of the project's behavior is handled by a state machine that advances through various states as the player plays content, preroll ads, midroll ads, and postroll ads. The initial state machine was designed to implement a specific set of functionality when the project was first conceptualized. Over the years, the project became much more fully-featured and many new scenarios and edge-cases arose. In this refactor, the state machine was updated to precisely match the modern feature-set. This has resulted in a more maintainable and reliable codebase.
+
+## How it works
+
+contrib-ads has a concept called _ad mode_. Ad mode is [strictly defined](https://github.com/videojs/videojs-contrib-ads#ad-mode-definition) as _if content playback is blocked by the ad plugin_. This does not necessarily mean that an ad is playing. For example, if there is a preroll ad, after a play request we show a loading spinner until ad playback begins. This is considered part of ad mode. The public method `player.ads.isInAdMode()` can be used to check if we are in ad mode.
```
Comment (gkatsev) [inline source/_drafts/videojs-contrib-ads-6.md]: Thoughts on making this sentence `Ad mode is [strictly defined](https://github.com/videojs/videojs-contrib-ads#ad-mode-definition) as _if content playback is blocked by the ad plugin_.` into a pullquote?
```ejs
{% pullquote right %}
Ad mode is [strictly defined](https://github.com/videojs/videojs-contrib-ads#ad-mode-definition) as _if content playback is blocked by the ad plugin_.
{% endpullquote %}

Comment (gkatsev) [reply source/_drafts/videojs-contrib-ads-6.md]: that is, in addition to having it inline with the paragraph.

File: source/_drafts/videojs-contrib-ads-6.md [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+In version 6, the state machine was refactored to match this strict definition of ad mode. There are two types of states: content states and ad states. If contrib-ads is in an ad state, it is in ad mode.
+
+Here is a diagram of the new states. Blue states are content states and yellow states are ad states.
+
+![](./ad-states.png)
+
+For a history of how the state machine has evolved over time, [there is detailed information in this Github issue](https://github.com/videojs/videojs-contrib-ads/issues/320). In this article we're going to focus on how it works in version 6.
+
+Let's walk through the states for a simple preroll scenario. The plugin begins in BeforePreroll state. This is considered a content state because content playback hasn't been requested yet and so ads are not blocking playback. However, the ad plugin can still asynchronously request ads from an ad server during this time. The ad plugin triggers the `adsready` event to say that it's ready. Contrib-ads knows that once the play button is pressed, ads are already prepared.
```
Comment (ldayananda) [inline source/_drafts/videojs-contrib-ads-6.md]: To improve clarity: refer to contrib-ads only by name rather than "the plugin" and an integrating ad plugin only as "ad plugin". I think this is the only place in the example that "the plugin" is used.

**Review by ldayananda — COMMENTED** [review]
One last suggestion, otherwise lgtm

File: source/_drafts/videojs-contrib-ads-6.md (hunk 9 above)
Comment (incompl) [reply source/_drafts/videojs-contrib-ads-6.md]: Will change

**Comment by incompl** [comment]
@gkatsev @ldayananda Did final review updates, plus a couple last final tweaks. I think it's ready to go.

**Review by gkatsev — APPROVED** [review]
Looking good https://deploy-preview-26--videojs-blog.netlify.com/
